Deprecate `_field_names` disabling

cbuescher opened the pull request:
Currently we allow `_field_names` fields to be disabled explicitely, but since
the overhead is negligible now we decided to keep it turned on by default and
deprecate and ignore the `enable` option on the field type.

Closes #27239

Commit:
Deprecate and ignore `_field_names` disabling

Currently we allow `_field_names` fields to be disabled explicitely, but since
the overhead is negligible now we decided to keep it turned on by default and
deprecate and ignore the `enable` option on the field type.

Closes #27239

elasticmachine commented:
Pinging @elastic/es-search

jtibshirani reviewed (commented):
Thanks @cbuescher, it will be very nice to remove this piece of configuration. I had a couple high-level questions:
- Am I understanding correctly that we are planning to deprecate the option and also start ignoring it in 7.x? I was wondering if we should follow the standard deprecation + removal process here, so that the configuration option will emit a deprecation warning in 7.x, but still work. I see that this PR takes the approach described in the original issue (#27239), but to me it seems most consistent + least surprising to users to follow the normal deprecation process.
- As it currently stands, I think we will log a warning on every mapping update (although it gets de-duplicated through `deprecatedAndMaybeLog`). This is because types are parsed every time we update mappings. If a user has specified the `enabled` setting, it doesn't seem like they have a way of addressing the warning unless they create a fresh mapping + reindex. I wanted to check that we're okay with this behavior.

jimczi commented on modules/percolator/src/test/java/org/elasticsearch/percolator/PercolatorFieldMapperTests.java:
```diff
@@ -146,7 +146,7 @@ public void init() throws Exception {
 
         String mapper = Strings.toString(XContentFactory.jsonBuilder().startObject().startObject("doc")
-            .startObject("_field_names").field("enabled", false).endObject() // makes testing easier
+            .startObject("_field_names").endObject() // makes testing easier
```
+1, we should disallow `_field_names` entirely for newly created indices.

cbuescher replied:
I'm not really following you here, #27239 talks about disabeling the `enable`option. What would disallowing the whole field mean? Do we still create the field (since its default is enabled:true?) What's the whole breaking/deprecation story around this then?

jimczi replied:
Sorry I wasn't clear I meant that we should disallow setting `_field_names`  entirely since we're on master but it seems that the scope of this pr is only the deprecation. However I agree with Julie that we should try to avoid mentioning `_field_names` in this test since this shouldn't affect the behavior of the percolator. If it turns too complicated we can leave this for a follow up but we'll need to do it anyway since the goal is to disallow disallowing the `_field_names` entirely and this test relies on this forbidden behavior.

jimczi commented on server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java:
```diff
@@ -105,7 +103,9 @@ public FieldNamesFieldMapper build(BuilderContext context) {
+                    deprecationLogger.deprecatedAndMaybeLog("field_names_enabled_parameter", ENABLED_DEPRECATION_MESSAGE);
+                    // just read the value and dump it on the floor
+                    XContentMapValues.nodeBooleanValue(fieldNode, name + ".enabled");
                     iterator.remove();
```
I think we should still support this field (and log the deprecation) for indices created in a version where the option is available (7x) and throw an error if the version is on or after 8.0 ?

jimczi reviewed (changes requested):
Thanks @cbuescher . I left some additional comments.

Commit:
iter

cbuescher commented:
Thanks for the review, I pushed a commit addressing the nits and typos.

>  understanding correctly that we are planning to deprecate the option and also start ignoring it in 7.x

I was wondering the same but since this PR is against 8.0 I was hoping we can discuss this here. Seeing that the scope of the issue now seems to shift slightly I think we should map out the real goals of #27239 before I proceed.

jimczi commented:
> I was wondering the same but since this PR is against 8.0 I was hoping we can discuss this here. Seeing that the scope of the issue now seems to shift slightly I think we should map out the real goals of #27239 before I proceed.

I think that most of the user of this feature (disallowing `_field_names` to index faster) are not aware of #26930 so they continue to disable it even though their mappings don't need to index anything in the `_field_names` field. However I agree with Julie that it would be least surprising + most consistent to only deprecate and not ignore especially for old indices. 

> As it currently stands, I think we will log a warning on every mapping update (although it gets de-duplicated through deprecatedAndMaybeLog). This is because types are parsed every time we update mappings. If a user has specified the enabled setting, it doesn't seem like they have a way of addressing the warning unless they create a fresh mapping + reindex. I wanted to check that we're okay with this behavior.

I think it's ok but if we choose to not ignore the option for old indices we should add the deprecation only if `enable` is set to false. This way users can remove the warning when they update their mapping to reenable the `_field_names` field ?

Commit:
Merge branch 'master' into fix-27239

Commit:
Only deprecate and log

cbuescher commented:
@jtibshirani @jimczi picking this up again, I reduced the scope of this PR to only add deprecation logging with the intent to backport this to 7.4 so we already have deprecation in 7 and then follow up in a separate PR to either completely disallow it (throwing an error) in 8 (only on newly created indices, the ones created in 7 should continue working) or just changing to a no-op in 8 and complete removal in a later version. Since this might also require some additons to the upgrade assistant I'd like to get deprecation in now and work on ignoring or removing the option entirely in a follow up PR. Wdyt?

jtibshirani commented:
It sounds good to me to focus only on deprecation for this PR. To make sure we're on the same page before jumping into the code, what do you both think of this high-level plan?
- On 7.4 we emit a deprecation warning if the mapping contains an explicit setting for 'enabled' (either true or false). We will emit this deprecation warning on 7.4 no matter when the index was created (6.x or 7.x).
- On 8.0 we disallow setting `enabled` on new indices. The setting continues to work for indices created in 7.x, but we still emit a deprecation warning. (Or just ignore the setting on 7.x, we can discuss this in the follow-up PR).

To me this seems like the most straightforward + least surprising plan. But it is a little annoying to see a deprecation warning every time the mapping is parsed, without an easy way to fix it without creating a fresh mapping + reindexing.  @jimczi suggested that we could only log a warning if `enabled: false`, but I was concerned that this would encourage users to set `enabled: true` without reindexing. Their indices would be in a confusing in-between state where `enabled: true` but some documents are missing `_field_names` information, so `exists` queries will return unexpected results.

Commit:
Merge branch 'master' into fix-27239

Commit:
Log deprecation on any index version, inlcude name in log

cbuescher commented:
@jtibshirani I'm in favour of the plan you proposed. I changed the PR slightly so that in its current form we emit a deprecation warning if the mapping contains an explicit setting for 'enabled' regardless of the value and the index version. I'm not too concerned with too many deprecation warnings. As far as I understand, warnings should be emitted on index creation, index opening  or mapping updates (which might happen when new fields are added either via explicit mapping updates or indexing new documents with new fields). All those cases should be limited to a few times, the warnings in the request header that are always there can be ignored by the client and the warnings in the log should largely be de-duplicated. 
I added information about the index that contains the "enabled" setting to make the deprecation warning more actionable.
It would be great to get the deprecation in with 7.4 but lets hear @jimczi opinion first.

Commit:
iter

Commit:
Merge branch 'master' into fix-27239

Commit:
Merge branch 'master' into fix-27239

cbuescher commented:
@jtibshirani @jimczi just fixed some conflicts on this PR, looks like FF for 7.4 already passed so I think the deprecation logging can be added for 7.5 next. Could you take another look and let me know if this reflects [Julies previous suggestion](https://github.com/elastic/elasticsearch/pull/42854#issuecomment-518400232)?
I would like to merge and backport this PR and open a follow up issue for the second point (throwing error on 8.0 on new indices), if you agree.

jtibshirani reviewed (approved):
This looks good to me as the initial deprecation PR.

cbuescher commented on server/src/main/java/org/elasticsearch/index/mapper/FieldNamesFieldMapper.java:
```diff
@@ -95,6 +95,11 @@ public FieldNamesFieldMapper build(BuilderContext context) {
 
     public static class TypeParser implements MetadataFieldMapper.TypeParser {
+
+        public static final String ENABLED_DEPRECATION_MESSAGE = "Index [{}] uses the `enabled` setting for `_field_names`, which is no "
```
I changed this, but left the later part that asks for removal of the setting in mappings and templates.

Commit:
Changes to deprecation message and revert removing test

cbuescher commented:
Thanks @jtibshirani, also checked with @jimczi on another channel who approves the plan as well. Will merge then and follow up with another PR on master to disallow the setting on new indices.
